statistics/semantic/src/main/java/com/spotify/heroic/statistics/semantic/SemanticMetricBackendReporter.java

```diff
@@ -147,16 +175,35 @@ public void reportDataNoLongerNeeded(final long n) {
             public void reportOperationEnded() {
                 // Decrement any remaining live memory
                 sampleSizeLive.dec(dataInMemory.get());
+
+                final long dataReadLocal = dataRead.get();
                 /* The accumulated count has previously been incremented with all of the data read,
                  * so decrease the full amount now */
-                sampleSizeAccumulated.dec(dataRead.get());
+                sampleSizeAccumulated.dec(dataReadLocal);
 
-                querySamplesRead.update(dataRead.get());
+                querySamplesRead.update(dataReadLocal);
                 queryRowsAccessed.update(rowsAccessed.get());
+                queryMaxLiveSamples.update(maxDataInMemory.get());
+
+                long duration = queryWatch.elapsed(TimeUnit.NANOSECONDS);
+                if (duration != 0) {
+                    queryReadRate.update((1000_000_000 * dataReadLocal) / duration);
+                }
+
+                for (ReadRangeStatistics stats : seriesDensity.values()) {
+                    long rangeMs = stats.getRange().diff();
+                    queryRowDensity.update((3600_000L * stats.getValue().get()) / rangeMs);
```

Comment:
Thanks for your comments, I've already implemented your suggestion @udoprog - but haven't pushed that update yet since I'm also working on exposing the data in query log.